### PR TITLE
Introduce GN target for Storybook

### PR DESCRIPTION
**Pull request by petemill**
Intention is to be able to build the deps only required which JS compiled by storybook uses.

Also allows gn target to be specified with --target flag, e.g. `npm run build -- --use_goma --target=brave:storybook`

I decided not to add this as a prefix to the current `npm run build-storybook` command since I didn't want to run the risk of causing developers a full build if e.g. the `--use_goma` flag wasn't included. When we have more reliable goma/no-goma builds, we can introduce this.

Resolves https://github.com/brave/brave-browser/issues/18551
Resolves https://github.com/brave/brave-browser/issues/16124
This will be followed-up with a change in the build scripts to run this build target before generating storybook.

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

**Comment by simonhong** [comment]
I have PR(https://github.com/brave/brave-core/pull/8959) for separated build target long times ago.
I'll close my PR :)

File: components/brave_vpn/BUILD.gn [hunk 1]
```diff
@@ -10,9 +10,8 @@ import("//tools/grit/preprocess_if_expr.gni")
 preprocess_folder = "preprocessed"
 preprocess_mojo_manifest = "preprocessed_mojo_manifest.json"
 
-assert(enable_brave_vpn)
-
 static_library("brave_vpn") {
+  assert(enable_brave_vpn)
```
Comment (bridiver) [inline components/brave_vpn/BUILD.gn:14]: why this change?

Comment (petemill) [reply components/brave_vpn/BUILD.gn:14]: `enabled_brave_vpn` is false in certain configurations (linux at the moment), but this build file is used for JS-only compiles as of this PR. So I moved the assert from the build-file-level to the browser c++ build level.

File: .storybook/BUILD.gn [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+group(".storybook") {
+  # Depend on any generated JS output that will
+  # be imported by any JS code used in any Brave
+  # storybook JS.
+  # Explicitly defined here so that even when those targets
+  # are disabled in a regular brave build due to build flags,
+  # they will be generated before storybook is compiled.
+  deps = [
+    "//brave/components/brave_vpn:mojom_js",
```
Comment (petemill) [inline .storybook/BUILD.gn:14]: I'm going to remove this line for now as it's breaking CI for linux without the devops PR being merged yet which would call `npm run build -- --target=brave:storybook` before running `npm run build-storybook` on every platform build.

**Review by bridiver — APPROVED** [review]

**Comment by petemill** [comment]
@simonhong thanks for the prior work - I didn't know about it!

**Review by simonhong — APPROVED** [review]
